src/PlatformEngines.jl

```diff
@@ -604,27 +604,70 @@ function get_server_dir(url::AbstractString)
     joinpath(depots1(), "servers", m.captures[1])
 end
 
+const AUTH_ERROR_HANDLERS = []
+
+function handle_auth_error(url, err; verbose::Bool = false)
+    handled, should_retry = false, false
+    for handler in AUTH_ERROR_HANDLERS
+        handled, should_retry = handler(url, Pkg.pkg_server(), err)
+        handled && break
+    end
+    handled && should_retry && return get_auth_header(url; verbose = verbose)
+    return nothing
+end
+
+"""
+    register_auth_error_handler(f)
+
+Registers `f` as the topmost handler for failures in package server authentication.
+
+`f` must be a function that takes three input arguments `(url, pkgserver, err)`, where `url` is the
+URL currently being downloaded, `pkgserver = Pkg.pkg_server()` the current package server, and
+`err` is one of `no-auth-file`, `insecure-connection`, `malformed-file`, `no-access-token`,
+`no-refresh-key`, `insecure-refresh-url`,  `malformed-file`, or `no-access-token`.
+
+The handler `f` needs to return a tuple of `Bool`s `(handled, should_retry)`. If `handled` is `false`,
+the next handler in the stack will be called, otherwise `get_auth_header` is called again if `should_retry`
+is `true`.
+
+`register_auth_error_handler` returns a zero-arg function that can be called to deregister the handler.
+"""
+function register_auth_error_handler(f)
+    unique!(pushfirst!(AUTH_ERROR_HANDLERS, f))
+    return () -> deregister_auth_error_handler(f)
+end
+
+"""
+    deregister_auth_error_handler(f)
+
+Removes `f` from the stack of authentication error handlers.
+"""
+function deregister_auth_error_handler(f)
+    filter!(handler -> handler !== f, AUTH_ERROR_HANDLERS)
+    return nothing
+end
+
 function get_auth_header(url::AbstractString; verbose::Bool = false)
     server_dir = get_server_dir(url)
-    server_dir === nothing && return
+    server_dir === nothing && return handle_auth_error(url, "no-auth-file"; verbose=verbose)
```

Comment:
Slight problem here: this should not call `handle_auth_error` when the request doesn't need authentication, i.e. when `pkg_server() == nothing` or when `!startswith(url, server)`. A few ways I can think to handle it:
1. factor the part where `get_server_dir` checks those two things out and handle them separately here;
2. change `get_server_dir` to return `Union{String, Bool}` where `false` means "this request doesn't need authentication" and `true` means "this request needs authentication but doesn't have an auth directory yet";
3. change `handle_auth_error` to check `pkg_server()` and `startswith(url, server)` again an bail out if the request doesn't need to be authenticated.

Unfortunately, they're all a bit messy.

Reply:
My thought was that whatever handler is registered also has the responsibility to check whether it's applicable (e.g. by checking the `Pkg.pkg_server()` argument).

Reply:
That seems like something that should be done once, not by each handler. If you're not talking to the current `pkg_server` then you shouldn't be sending auth headers at all.

Reply:
Fair enough. Imho the best solution would be to register a pair of `Union{Regex, String}, Function` and check whether any given handler is applicable with `occursin`.

Reply:
Should be resolved with the current design.

Reply:
Was resolved by reverting the first `return` not to call any handlers.